src/protocol/attribution/apply_attribution_window.rs

```diff
@@ -0,0 +1,351 @@
+        .take_while(|&v| v < num_rows)
+        .enumerate()
+    {
+        let record_count = num_rows - step_size;
+
+        let depth_i_ctx = ctx
+            .narrow(&InteractionPatternStep::from(depth))
+            .set_total_records(record_count);
+        let mt_ctx = depth_i_ctx.narrow(&Step::PreviousStopBitTimesCurrent);
+
+        let mut futures = Vec::with_capacity(record_count);
+
+        // `i` starts from 0 for RecordId. The actual logic skips the first
+        // `step_size` rows.
+        for i in 0..record_count {
+            let index = i + step_size;
+
+            let c1 = depth_i_ctx.clone();
+            let c2 = mt_ctx.clone();
+            let record_id = RecordId::from(i);
+
+            let previous_stop_bit = &stop_bits[index - step_size];
+            let current_stop_bit = &stop_bits[index];
+            let previous_t_delta = &t_delta[index - step_size];
+
+            futures.push(async move {
+                try_join(
+                    // if the current stop_bit is 0, don't accumulate
+                    c1.multiply(record_id, previous_t_delta, current_stop_bit),
+                    // next `stop_bit`
+                    c2.multiply(record_id, previous_stop_bit, current_stop_bit),
+                )
+                .await
+            });
+        }
+
+        let results = try_join_all(futures).await?;
+
+        results
+            .into_iter()
+            .enumerate()
+            .for_each(|(i, (delta, stop_bit))| {
+                t_delta[i + step_size] += &delta;
+                stop_bits[i + step_size] = stop_bit;
+            });
+    }
+
+    Ok(t_delta)
+}
+
+/// Creates a vector of trigger values where values are set to `0` if the time delta
+/// from their nearest source event exceed the specified attribution window cap.
+///
+/// This protocol executes the bit-decomposition protocol in order to compare shares
+/// of time deltas in `F`.
+///
+/// # Errors
+/// Fails if the bit-decomposition, bitwise comparison, or multiplication fails.
+async fn zero_out_expired_trigger_values<F, C, T>(
+    ctx: &C,
+    input: &[MCApplyAttributionWindowInputRow<F, T>],
+    time_delta: &[T],
+    cap: u32,
+) -> Result<Vec<T>, Error>
+where
+    F: Field,
+    C: Context<F, Share = T>,
+    T: Arithmetic<F>,
+{
+    // Compare the accumulated timestamp deltas with the specified attribution window
+    // cap value, and zero-out trigger event values that exceed the cap.
+    let c = ctx.clone().set_total_records(input.len());
+    let bit_decomposition_ctx = c.narrow(&Step::TimeDeltaBitDecomposition);
+    let cmp_ctx = c.narrow(&Step::TimeDeltaLessThanCap);
+    let mul_ctx = c.narrow(&Step::CompareBitTimesTriggerValue);
+
+    let random_bits_generator =
+        RandomBitsGenerator::new(ctx.narrow(&Step::RandomBitsForBitDecomposition));
+    let rbg = &random_bits_generator;
+
+    try_join_all(
+        zip(input, time_delta)
```

Comment:
I think you can use input.iter().skip(1) or something to skip the first row because:
- It's either a source or a trigger report
- If it's a source report, then the result will be zero due to the final multiplication
- if it's a trigger report, then there is no way there is any source event before it to which it could have been attributed.

Reply:
...but this is a *micro* optimization. So feel free to do this in some later diff.

Reply:
I'll follow up in a later diff.